### PR TITLE
Implement JSON.Encoder for Calendar types

### PR DESCRIPTION
**Pull request by josevalim**
Should be backported if accepted.

### CODE REVIEW

**Comment by sorentwo** [comment]
Really pleased to see this. It was the only barrier to a clean transition from Jason for my benchmarking purposes.

**Comment by josevalim** [comment]
Can you manually add it to your benches for now and let us know the numbers? :)

**Comment by sorentwo** [comment]
> Can you manually add it to your benches for now and let us know the numbers? :)

Sure thing. After applying the patch the test suites all pass, and here's the benchmark:

```elixir
args = %{
  a: 1,
  b: "value",
  c: true,
  d: "baz",
  e: [1],
  f: %{f1: 123},
  g: [%{g1: "bat"}],
  h: "abc-123",
  i: ["abc-123"],
  j: ["foo", "bar"],
  k: [:a, :b],
  l: Date.utc_today(),
  m: Time.utc_now(),
  n: DateTime.utc_now(),
  o: NaiveDateTime.utc_now()
}

recode = fn encoder ->
  args
  |> encoder.encode!()
  |> encoder.decode!()
end

Benchee.run(%{"Encode/Decode" => recode}, inputs: Map.new([Jason, JSON], &{&1, &1}))
```

And here are the results:

```
Operating System: macOS
CPU Information: Apple M3 Pro
Number of Available Cores: 12
Available memory: 36 GB
Elixir 1.18.0-dev
Erlang 27.0
JIT enabled: true

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 5 s
memory time: 0 ns
reduction time: 0 ns
parallel: 1
inputs: Elixir.JSON, Elixir.Jason
Estimated total run time: 14 s

Benchmarking Encode/Decode with input Elixir.JSON ...
Benchmarking Encode/Decode with input Elixir.Jason ...
Calculating statistics...
Formatting results...

##### With input Elixir.JSON #####
Name                    ips        average  deviation         median         99th %
Encode/Decode      175.89 K        5.69 μs   ±219.43%        4.67 μs       15.88 μs

##### With input Elixir.Jason #####
Name                    ips        average  deviation         median         99th %
Encode/Decode      150.05 K        6.66 μs   ±169.02%        5.08 μs       21.13 μs
```

It's a great improvement for a synthetic, small, but type exhaustive benchmark.

**Review by sabiwara — APPROVED** [review]
LGTM!

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

File: lib/elixir/lib/json.ex [hunk 1]
```diff
@@ -150,6 +150,12 @@ defimpl JSON.Encoder, for: Map do
   end
 end
 
+defimpl JSON.Encoder, for: [Date, Time, NaiveDateTime, DateTime] do
```
Comment (Wigny) [inline lib/elixir/lib/json.ex:153]: Should `Duration` also be included?

Comment (josevalim) [reply lib/elixir/lib/json.ex:153]: I asked myself the same and, honestly,  I am not sure. We could easily use the ISO representation, but I am not sure how widespread it is, it is definitely less common than the datetime ones.

Comment (Wigny) [reply lib/elixir/lib/json.ex:153]: AFAIK, in the JS world it is common to serialise duration to ISO strings when parsing it to JSON (see [`Temporal.Duration.toJSON`](https://tc39.es/proposal-temporal/docs/duration.html#toJSON) or [`dayjs.duration`](https://day.js.org/docs/en/durations/as-json)). As the string format chose in serialization is just a convention, I hopped ISO strings was a good enough format to be used as default.

**Pull request merged**